shell/platform/android/io/flutter/plugin/platform/SingleViewPresentation.java

```diff
@@ -154,7 +170,7 @@ protected void onCreate(Bundle savedInstanceState) {
         }
 
         container = new FrameLayout(getContext());
-        PresentationContext context = new PresentationContext(getContext(), state.windowManagerHandler);
+        Context context = wrapContext(getContext(), state.windowManagerHandler);
```

Comment:
I find this somewhat confusing, my thought process was: wrapContext is going to replace the IMM with the one it's getting from the context I'm giving it, but we're giving it the inner context which has the wrong IMM... ohh the innerContext is already wrapped at this point.

The other thing I'm finding not obvious from the code is that `getContext()` is already a "wrapped context", without digging into `Presentation` and `ContextWrapper` it's not obvious why we do this here.

I'm also a little worried about being confused by stack traces with 2 separate instances of PresentationContext.

I guess some of this can be mitigated with comments, alternatively we can implement our own Presentation and have a single context wrapper.

Reply:
Tried to make this clearer in the last patch.

I added a comment, which is a way to clear this up while reading the code but wouldn't help with a debugger. 

I also split the two main wrappers (IMM at construction time, both IMM and WindowManager onCreate) into two separate classes. It's still murky but I _think_ this will make it clearer when going through this with a debugger since they'll show up as different classes. I experimented a little with trying to cast `getContext()` here into something intelligible, but because `Presentation` is wrapping it there wasn't anything I could really do.

It occurred to me that I could also extend Presentation and fork the constructor and `createPresentationContext` methods instead of just outright forking the entire class, which I think would mitigate most of my concerns with maintaining a fork of it.

WDYT?